// react-ctx/lib/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var type = exports.type = function type(x) {
  return {}.toString.call(x).slice(8, -1);
};

var pick = exports.pick = function pick(original) {
  var keys = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  return keys.reduce(function (created, key) {
    if (key in original) {
      created[key] = original[key];
    }

    return created;
  }, {});
};

var omit = exports.omit = function omit(original) {
  var keys = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  return Object.keys(original).reduce(function (created, key) {
    if (keys.indexOf(key) === -1) {
      created[key] = original[key];
    }

    return created;
  }, {});
};

var mapValues = exports.mapValues = function mapValues(original, f) {
  return Object.keys(original).reduce(function (created, key) {
    created[key] = f(original[key]);
    return created;
  }, {});
};

var mapKeys = exports.mapKeys = function mapKeys(original, f) {
  return Object.keys(original).reduce(function (created, key) {
    created[f(key)] = original[key];
    return created;
  }, {});
};

var keyMap = exports.keyMap = function keyMap(keys) {
  return keys.reduce(function (created, key) {
    created[key] = key;

    return created;
  }, {});
};
